src/strings.rs

```diff
@@ -0,0 +1,211 @@
+///
+/// ```python
+/// numpy.array(["foo", "bar", "foobar"])
+/// ```
+///
+/// yielding `U6`, NumPy will automatically determine the smallest possible array length at runtime.
+/// This does not work well with the length `N` of `PyUSC4Array<N>` determined at compile-time
+/// and hence explicitly setting the `dtype`, e.g.
+///
+/// ```python
+/// numpy.array(["foo", "bar", "foobar"], dtype='U12')
+/// ```
+///
+/// always matching `PyArray1<PyUSC4Array<12>>`, is recommended.
+///
+/// # Example
+///
+/// ```rust
+/// # use pyo3::Python;
+/// use numpy::{PyArray1, PyFixedUnicode};
+///
+/// # Python::with_gil(|py| {
+/// let array = PyArray1::<PyFixedUnicode<3>>::from_vec(py, vec![[b'b' as _, b'a' as _, b'r' as _].into()]);
+///
+/// assert!(array.dtype().to_string().contains("U3"));
+/// # });
+/// ```
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PyFixedUnicode<const N: usize>(pub [Py_UCS4; N]);
+
+impl<const N: usize> fmt::Display for PyFixedUnicode<N> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        for character in self.0 {
+            if character == 0 {
+                break;
+            }
+
+            write!(fmt, "{}", char::from_u32(character).unwrap())?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<const N: usize> From<[Py_UCS4; N]> for PyFixedUnicode<N> {
+    fn from(val: [Py_UCS4; N]) -> Self {
+        Self(val)
+    }
+}
+
+unsafe impl<const N: usize> Element for PyFixedUnicode<N> {
+    const IS_COPY: bool = true;
+
+    fn get_dtype(py: Python) -> &PyArrayDescr {
+        static DTYPES: TypeDescriptors = TypeDescriptors::new(NPY_TYPES::NPY_UNICODE, b'=' as _);
+
+        unsafe { DTYPES.from_size(py, size_of::<Self>()) }
+    }
+}
+
+struct TypeDescriptors {
+    npy_type: NPY_TYPES,
+    byteorder: c_char,
+    #[allow(clippy::type_complexity)]
+    dtypes: GILProtected<RefCell<Option<FxHashMap<usize, Py<PyArrayDescr>>>>>,
+}
+
+impl TypeDescriptors {
+    const fn new(npy_type: NPY_TYPES, byteorder: c_char) -> Self {
+        Self {
+            npy_type,
+            byteorder,
+            dtypes: GILProtected::new(RefCell::new(None)),
+        }
+    }
+
+    /// `self.npy_type` must be either `NPY_STRING` or `NPY_UNICODE`
+    /// with matching byteorder and `size` must match the element size in bytes
+    #[allow(clippy::wrong_self_convention)]
+    unsafe fn from_size<'py>(&'py self, py: Python<'py>, size: usize) -> &'py PyArrayDescr {
+        let mut dtypes = self.dtypes.get(py).borrow_mut();
+
+        let dtype = match dtypes.get_or_insert_with(Default::default).entry(size) {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => {
+                let dtype = PyArrayDescr::new_from_npy_type(py, self.npy_type);
+
+                let descr = &mut *dtype.as_dtype_ptr();
+                descr.elsize = size.try_into().unwrap();
```

Comment:
Can't we use `as`?

Reply:
We could but I would like to panic here in case of overflow (`size: usize` will usually be much larger than `elsize: i32`) as otherwise memory unsafety could ensue.